### PR TITLE
16 create significant strike features

### PR DESCRIPTION
**Pull request by straw-hat-gupta**

### CODE REVIEW

File: significant_strikes.py
Comment (ivan-guo) [inline significant_strikes.py:37]: This might be a better way to do it rather than directly inserting the new columns into the original frame via the function - I don't remember but I think I did this for some functions in my Tennis model does this get rid of the warnings when running it? If it does we might have to change all the other issues to also create features this way.

Comment (straw-hat-gupta) [reply significant_strikes.py:37]: Yeah it gets rid of the PerformanceWarning and the SettingWithCopyWarning.

Comment (ivan-guo) [reply significant_strikes.py:37]: 🔥

File: significant_strikes.py
Comment (ivan-guo) [inline significant_strikes.py:218]: A more concise way to do this would be:

`return prev_fights[-3:], prev_fights[-5:], prev_fights`

This gets the last X fights as required but also gets the entire dataset if there are less than X rows.

File: significant_strikes.py
Comment (ivan-guo) [inline significant_strikes.py:258]: String comparisons are costly and as a general coding practice tip, it's better to keep a variable as one type rather than multiple for error handling/debugging purposes. I think in this case it's fine, (you could just make round '0' for the overall case) since it does make the code a bit more readable.

File: significant_strikes.py
Comment (ivan-guo) [inline significant_strikes.py:283]: For this for loop I think you can make it faster and more concise:

`stats_df = pd.concat([fight.apply(lambda row: self.get_round_time(row['outcome_round'], row['outcome_time'], round), fight['fighter_a_id], fight['fighter_b_id'], [fight[col] for col in sig_str_cols], ], keys=['total_fight_time', fighter_a_id', 'fighter_b_id', ...], axis=1)`

I'm not entirely sure of the syntax, but you can definitely vectorize this operation and concat an apply function with other vectorization techniques to make this much faster. Speed is definitely important when it comes to this feature set because it's a very large set.

File: significant_strikes.py
Comment (ivan-guo) [inline significant_strikes.py:290]: Don't forget that the fighter_a in one fight could be fighter_b in the next, so make sure that you're grabbing their stats by a or b and not that fighter_a is getting all the fighter_a stats in the last X fights regardless of whether they are fighter_a in every fight or not

File: significant_strikes.py
Comment (ivan-guo) [inline significant_strikes.py:317]: You can use the .sum() function to sum up columns. In this case it is a little tricky, but you can do something like this:

`fighter_a_id_vals = stats_by_fight_df.fighter_a_id.values`
`fighter_b_id_vals = stats_by_fight_df.fighter_b_id.values`
`sig_str_landed = stats_by_fight_df[fighter_a_id_vals == fighter_id]['fighter_a_sig_str_landed'].sum() + stats_by_fight_df[fighter_b_id_vals == fighter_id]['fighter_b_sig_str_landed'].sum()`

This will vectorize the operation and make it MUCH faster than doing a for loop. 
See lines 53-57 [here](https://github.com/ivan-guo/OctaPicks/pull/33/commits/eb2ee116f23f5cac5c4668735009b88385c99f01).

File: significant_strikes.py
Comment (ivan-guo) [inline significant_strikes.py:500]: I think we should start writing some tests to assert that the dataset is actually correct. Test sample shouldn't be the entire dataset but what I think we could do is take one fighter, index the dataset into just that one fighter's fights, and see how the features evolve for that specific fighter over there fights and check validity from there. You don't have to write the test now but it will get more difficult as we make a bunch of features and if it doesn't work it can be hard to see where it broke at.

File: significant_strikes.py
Comment (straw-hat-gupta) [reply significant_strikes.py:290]: yah thats why i get the colomns for both fighters a and b then in the other functions i get the correct fighter's stat based on fighter_id

Comment (ivan-guo) [reply significant_strikes.py:290]: okok fire

File: significant_strikes.py
Comment (straw-hat-gupta) [reply significant_strikes.py:500]: I agree. i was gonna write tests for my features once you wrote your tests to see how you did it lol.

**Comment by ivan-guo** [comment]
This is really good though you have good formatting and applying the function to a copy rather than direct slice is something we should carry over to other issues, but I think speed is definitely the biggest factor here and I think some parts of the code can be sped up. Otherwise this feature might take about ~10 hours on estimate due to size of our dataset and number of columns.

File: significant_strikes.py
Comment (ivan-guo) [reply significant_strikes.py:500]: Word I was have trouble because of the fighter_a not being the same fighter_a issue, because I want to apply the df.rolling() function [(here)](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rolling.html) so that I can get the rolling data and verify it, but not being the same fighter columns every fight makes it difficult. I think I'm going to swap every fighter and their respective columns so that the same fighter is in the fighter_a column then do the rolling from there.

File: significant_strikes.py
Comment (straw-hat-gupta) [reply significant_strikes.py:317]: ok bet imma apply this to the other funcs that do a similar thing

**Review by ivan-guo — CHANGES_REQUESTED** [review]
This is really good though you have good formatting and applying the function to a copy rather than direct slice is something we should carry over to other issues, but I think speed is definitely the biggest factor here and I think some parts of the code can be sped up. Otherwise this feature might take about ~10 hours on estimate due to size of our dataset and number of columns.

**Review by ivan-guo — APPROVED** [review]
lgtm, thanks for changing !

**Comment by ivan-guo** [comment]
You can merge this when you're ready - or wait until you finish a test

**Review by aryang13 — APPROVED** [review]

**Comment by straw-hat-gupta** [comment]
Don't merge yet. im gonna update a couple of things with this in the morning and merge it then.

**Pull request merged**
